classpath/java/io/BufferedInputStream.java

```diff
@@ -43,7 +43,12 @@ public int read() throws IOException {
 
   public int read(byte[] b, int offset, int length) throws IOException {
     int count = 0;
-
+    if (position >= limit) {
+	fill();
+      if (limit == -1) {
+        return -1;
+      }
+	}
```

Comment:
Oh, wait, sorry.  For some reason I was reading the implementation of the single-byte read() above as fill.  (facepalm).

With the _real_ fill implementation in mind, this is probably the best you can do.